### PR TITLE
Added Vector Summation Accumulator

### PR DESCRIPTION
**Pull request by levzlotnik**
## Description
Added `VectorSummationAccumulator` as per #43.  

## Note
Only implemented the summation part. The anonymization part will be implemented in a separate PR as requested by @dvadym .

## Affected Dependencies
None.

## How has this been tested?
Added some test cases in `tests/accumulator_tests.py` next to the generic ones.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: pipeline_dp/accumulator.py [hunk 1]
```diff
@@ -174,6 +185,47 @@ def compute_metrics(self) -> float:
...
+    noise_kind: NoiseKind
+    max_norm: float
+
+
+class VectorSummationAccumulator(Accumulator):
+    _vec_sum: np.ndarray
+
+    def __init__(self, params: VectorSummationParams,
+                 values: Iterable[Tuple[float]]) -> None:
+        self.sum_params = params
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:198]: Nit: _sum_params  (maybe simply _params)

File: pipeline_dp/accumulator.py [hunk 2]
```diff
@@ -174,6 +185,47 @@ def compute_metrics(self) -> float:
...
+        values = iter(values)
+        try:
+            self._vec_sum = np.array(next(values))
+            for val in values:
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:202]: values is already iterable, why not simply
`for val in values: `
w/o processing of StopIteration ?

We can assume that values is not empty (this object is created after GroupBy by key for each key which is present)

File: pipeline_dp/accumulator.py [hunk 3]
```diff
@@ -174,6 +185,47 @@ def compute_metrics(self) -> float:
...
+                self.add_value(val)
+        except StopIteration:
+            self._vec_sum = None
+
+    def add_value(self, value: Tuple[float]):
+        value = np.array(value)  # type: np.ndarray
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:208]: I'm thinking, whether np.array is better (less space, more effective?) than tuple for serializing and sending between workers, don't you know?

File: pipeline_dp/accumulator.py [hunk 4]
```diff
@@ -42,6 +46,13 @@ def add_value(self, value):
     """
         pass
 
+    def _check_accumulator(self, accumulator: 'Accumulator'):
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:49]: Maybe _check_is_mergeable? or _can_merge?

File: tests/accumulator_test.py [hunk 5]
```diff
@@ -198,9 +218,7 @@ def add_value(self, v):
 
     def add_accumulator(self,
                         accumulator: 'MeanAccumulator') -> 'MeanAccumulator':
-        if not isinstance(accumulator, MeanAccumulator):
-            raise TypeError(
-                "The accumulator to be added is not of the same type.")
+        self._check_accumulator(accumulator)
```
Comment (dvadym) [inline tests/accumulator_test.py:221]: Thanks for improving existing code!

**Review by dvadym — COMMENTED** [review]
Thanks it looks good. I left some comments and questions

File: pipeline_dp/accumulator.py (hunk 3 above)
Comment (levzlotnik) [reply pipeline_dp/accumulator.py:208]: `np.ndarray` is hands down better. Less memory, enjoys low level and hardware vectorization. I'm not certain if all serializations on it are as compact, but at the very least:
```python
import numpy as np
import pickle

# probably around 8000 byes of memory including metadata like dtype, shape and strides
x = np.arange(1000, dtype=np.int64)

x_p = pickle.dumps(x)
print(len(x_p)) # 8152

x_t = tuple(x) # same array but in tuple
x_t_p = pickle.dumps(x_t)
print(len(x_p)) # 19100
```
My assumption is that numpy is extremely optimized for computation and data movement, whether it's with or without multiprocessing.

File: tests/accumulator_test.py (hunk 5 above)
Comment (levzlotnik) [reply tests/accumulator_test.py:221]: No problem :)

File: pipeline_dp/accumulator.py (hunk 2 above)
Comment (levzlotnik) [reply pipeline_dp/accumulator.py:202]: In the tests we have empty accumulators like `XAccumulator(None, []).add_value(x)`.  
IMHO this is an edge case that must be treated in an expected manner. For example, a user that wrote some custom code for joining two tables and then grouped by some multi-key might make a mistake in implementation, and they should be able to get an indication of what went wrong.  
We can either leave the aggregated value as `None` or throw an error. @dvadym up to you :)

File: pipeline_dp/accumulator.py (hunk 1 above)
Comment (levzlotnik) [reply pipeline_dp/accumulator.py]: Done.

File: pipeline_dp/accumulator.py (hunk 4 above)
Comment (levzlotnik) [reply pipeline_dp/accumulator.py]: `_check_mergeable` is good enough?

File: pipeline_dp/accumulator.py (hunk 3 above)
Comment (dvadym) [reply pipeline_dp/accumulator.py:209]: Wow. Thanks for checking! Yeah, np arrays look much better

File: pipeline_dp/accumulator.py [hunk 6]
```diff
@@ -174,6 +186,48 @@ def compute_metrics(self) -> float:
...
+class VectorSummationParams:
+    noise_kind: NoiseKind
+    max_norm: float
+
+
+class VectorSummationAccumulator(Accumulator):
+    _vec_sum: np.ndarray
+
+    def __init__(self, params: VectorSummationParams,
+                 values: Iterable[Tuple[float]]) -> None:
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:198]: np.arrays are used very widely in Python and they are so much effective,  that it would be very ineffective for API clients which have data in np.arrays to convert to tuples. So I think it's worth to allow value to be not only tuple but np.array as well.

File: pipeline_dp/accumulator.py [hunk 7]
```diff
@@ -3,7 +3,11 @@
 import pickle
 from dataclasses import dataclass
 from functools import reduce
+
+from typing import Iterable, Tuple
 import pipeline_dp
+from pipeline_dp.aggregate_params import NoiseKind
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:9]: Nit: import only modules (according to [styleguide](https://google.github.io/styleguide/pyguide.html#22-imports) )

File: pipeline_dp/accumulator.py (hunk 6 above)
Comment (levzlotnik) [reply pipeline_dp/accumulator.py:198]: Even though it's only a type hint and the users are free to use `np.ndarray` - I agree with that point. We should allow users (or even soft-requiring them) to use `np.ndarray`.

File: pipeline_dp/accumulator.py (hunk 7 above)
Comment (levzlotnik) [reply pipeline_dp/accumulator.py]: Done

File: pipeline_dp/accumulator.py (hunk 6 above)
Comment (levzlotnik) [reply pipeline_dp/accumulator.py]: Done

File: tests/accumulator_test.py [hunk 8]
```diff
@@ -129,14 +129,34 @@ def test_merge_accumulators(self):
 
         self.assertEqual(merged_accumulator.compute_metrics(), 10)
 
+        vec_sum_accumulator1 = accumulator.VectorSummationAccumulator(
+            params=None,
+            values=[(15, 2)])
+        vec_sum_accumulator2 = accumulator.VectorSummationAccumulator(
+            params=None,
+            values=[(27, 40)])
```
Comment (dvadym) [inline tests/accumulator_test.py:137]: Please use np.array in some of the input, to test that np.arrays are processed correctly

File: pipeline_dp/accumulator.py [hunk 9]
```diff
@@ -174,6 +187,56 @@ def compute_metrics(self) -> float:
...
+    noise_kind: aggregate_params.NoiseKind
+    max_norm: float
+
+_FloatVector = Union[Tuple[float], np.ndarray]
+
+class VectorSummationAccumulator(Accumulator):
+    _vec_sum: np.ndarray
+
+    def __init__(self, params: VectorSummationParams,
+                 values: Iterable[_FloatVector]) -> None:
```
Comment (dvadym) [inline pipeline_dp/accumulator.py:200]: As mentioned in a comment below, let's assume that values is List (and then it would be possible to simplify the code)

**Review by dvadym — COMMENTED** [review]
Thanks, it looks great, just a few small comments

File: pipeline_dp/accumulator.py (hunk 9 above)
Comment (levzlotnik) [reply pipeline_dp/accumulator.py:199]: I just noticed that during `add_value` I check for `self._vec_sum is None` - I can easily just loop over `values` without assuming anything except that `values` is an iterable. No need to assume it's a `List`, also I think it's better to leave as `Iterable` as maybe in the future we'd like to do total aggregation over the entire dataset without group-by key.

File: pipeline_dp/accumulator.py [hunk 10]
```diff
@@ -3,7 +3,12 @@
 import pickle
 from dataclasses import dataclass
 from functools import reduce
+
+from typing import Iterable, Tuple, Union
 import pipeline_dp
+# from pipeline_dp.aggregate_params import NoiseKind
```
[the remark this answers is not in the document]
Comment (levzlotnik) [reply pipeline_dp/accumulator.py]: Done.

File: tests/accumulator_test.py (hunk 8 above)
Comment (levzlotnik) [reply tests/accumulator_test.py:139]: Done.

File: pipeline_dp/accumulator.py (hunk 9 above)
Comment (dvadym) [reply pipeline_dp/accumulator.py:199]: Ok, it makes sense! I agree we don't need to limit ourselves to list w/o necessity.

**Review by dvadym — APPROVED** [review]
Thanks!
